components/eam/src/physics/crm/crm_rad_module.F90

```diff
@@ -8,65 +8,105 @@ module crm_rad_module
 
-      rad%nc => null()
-      rad%ni => null()
-      rad%qs => null()
-      rad%ns => null()
+      call phys_getopts(MMF_microphysics_scheme_out = MMF_microphysics_scheme)
+
+      if (.not. allocated(rad%qrad))        allocate(rad%qrad       (ncrms, crm_nx_rad, crm_ny_rad, crm_nz))
+      if (.not. allocated(rad%temperature)) allocate(rad%temperature(ncrms, crm_nx_rad, crm_ny_rad, crm_nz))
+      if (.not. allocated(rad%qv))          allocate(rad%qv         (ncrms, crm_nx_rad, crm_ny_rad, crm_nz))
+      if (.not. allocated(rad%qc))          allocate(rad%qc         (ncrms, crm_nx_rad, crm_ny_rad, crm_nz))
+      if (.not. allocated(rad%qi))          allocate(rad%qi         (ncrms, crm_nx_rad, crm_ny_rad, crm_nz))
+      if (.not. allocated(rad%cld))         allocate(rad%cld        (ncrms, crm_nx_rad, crm_ny_rad, crm_nz))
+      
+      call prefetch(rad%qrad)
+      call prefetch(rad%temperature)
+      call prefetch(rad%qv)
+      call prefetch(rad%qc)
+      call prefetch(rad%qi)
+      call prefetch(rad%cld)
+
+      rad%qrad        = 0
+      rad%temperature = 0
+      rad%qv          = 0
+      rad%qc          = 0
+      rad%qi          = 0
+      rad%cld         = 0
+
+      if (MMF_microphysics_scheme .eq. 'm2005') then
+         if (.not. allocated(rad%nc)) allocate(rad%nc(ncrms, crm_nx_rad, crm_ny_rad, crm_nz))
+         if (.not. allocated(rad%ni)) allocate(rad%ni(ncrms, crm_nx_rad, crm_ny_rad, crm_nz))
+         if (.not. allocated(rad%qs)) allocate(rad%qs(ncrms, crm_nx_rad, crm_ny_rad, crm_nz))
+         if (.not. allocated(rad%ns)) allocate(rad%ns(ncrms, crm_nx_rad, crm_ny_rad, crm_nz))
+      
+         call prefetch(rad%nc)
+         call prefetch(rad%ni)
+         call prefetch(rad%qs)
+         call prefetch(rad%ns)
+      
+         rad%nc = 0
+         rad%ni = 0
+         rad%qs = 0
+         rad%ns = 0
+      end if
 
    end subroutine crm_rad_initialize
    !------------------------------------------------------------------------------------------------
    subroutine crm_rad_finalize(rad)
+      use phys_control, only: phys_getopts
       class(crm_rad_type), intent(inout) :: rad
 
-      ! Nullify pointers
-      rad%qrad => null()
-      rad%temperature => null()
-      rad%qv => null()
-      rad%qi => null()
-      rad%cld => null()
-
-      rad%nc => null()
-      rad%ni => null()
-      rad%qs => null()
-      rad%ns => null()
+      character(len=16) :: MMF_microphysics_scheme    ! CRM microphysics scheme
+
+      call phys_getopts(MMF_microphysics_scheme_out = MMF_microphysics_scheme)
+
+      if (allocated(rad%qrad))        deallocate(rad%qrad)
+      if (allocated(rad%temperature)) deallocate(rad%temperature)
+      if (allocated(rad%qv))          deallocate(rad%qv)
+      if (allocated(rad%qc))          deallocate(rad%qc)
+      if (allocated(rad%qi))          deallocate(rad%qi)
+      if (allocated(rad%cld))         deallocate(rad%cld)
+
+      if (MMF_microphysics_scheme .eq. 'm2005') then
```

Comment:
Just a note, I don't think this needs to be changed, but with the addition of the `if (allocated())` patterns, we don't actually need the query and logical about the microphysics scheme.

Reply:
yea, I figured we might as well skip the check of whether of not stuff is allocated, but I have no idea if there's any performance implication.

Reply:
No problem.